File: internal/tools/tools.go
// Package tools implements the read-only Kubernetes operations kubewhy can
// call: get, describe, logs, events, top. Every function here issues exactly
// one kind of API call (get/list/watch or a logs subresource read) -- none of
// them can mutate cluster state.
//
// That's not just a convention: kubewhy is meant to run as a ServiceAccount
// whose ClusterRole only grants get/list/watch (see
// deploy/readonly-clusterrole.yaml), so even a prompt-injected or
// hallucinating model gets a 403 from the API server if it tries anything
// else.
package tools

import (
	"context"
	"fmt"
	"sort"

	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/tools/clientcmd"
	metricsv "k8s.io/metrics/pkg/client/clientset/versioned"
)

type Client struct {
	core    kubernetes.Interface
	metrics metricsv.Interface
}

// Core exposes the underlying clientset for callers that need lower-level
// read-only access (e.g. the watcher's pod-listing poll loop).
func (c *Client) Core() kubernetes.Interface { return c.core }

func LoadClient() (*Client, error) {
	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
	config, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, &clientcmd.ConfigOverrides{}).ClientConfig()
	if err != nil {
		return nil, fmt.Errorf("loading kubeconfig: %w", err)
	}
	cs, err := kubernetes.NewForConfig(config)
	if err != nil {
		return nil, fmt.Errorf("building clientset: %w", err)
	}
	ms, err := metricsv.NewForConfig(config)
	if err != nil {
		return nil, fmt.Errorf("building metrics clientset: %w", err)
	}
	return &Client{core: cs, metrics: ms}, nil
}

type GetResourceInput struct {
	Kind          string `json:"kind"`
	Namespace     string `json:"namespace"`
	Name          string `json:"name,omitempty"`
	LabelSelector string `json:"label_selector,omitempty"`
}

func (c *Client) GetResource(ctx context.Context, in GetResourceInput) (any, error) {
	opts := metav1.ListOptions{LabelSelector: in.LabelSelector}
	switch in.Kind {
	case "pod":
		if in.Name != "" {
			return c.core.CoreV1().Pods(in.Namespace).Get(ctx, in.Name, metav1.GetOptions{})
		}
		return c.core.CoreV1().Pods(in.Namespace).List(ctx, opts)
	case "deployment":
		if in.Name != "" {
			return c.core.AppsV1().Deployments(in.Namespace).Get(ctx, in.Name, metav1.GetOptions{})
		}
		return c.core.AppsV1().Deployments(in.Namespace).List(ctx, opts)
	case "replicaset":
		return c.core.AppsV1().ReplicaSets(in.Namespace).List(ctx, opts)
	case "service":
		if in.Name != "" {
			return c.core.CoreV1().Services(in.Namespace).Get(ctx, in.Name, metav1.GetOptions{})
		}
		return c.core.CoreV1().Services(in.Namespace).List(ctx, metav1.ListOptions{})
	case "node":
		if in.Name != "" {
			return c.core.CoreV1().Nodes().Get(ctx, in.Name, metav1.GetOptions{})
		}
		return c.core.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
	default:
		return nil, fmt.Errorf("unsupported kind for get_resource: %s", in.Kind)
	}
}

type DescribePodInput struct {
	Namespace string `json:"namespace"`
	Name      string `json:"name"`
}

type DescribePodOutput struct {
	Pod    *corev1.Pod     `json:"pod"`
	Events []corev1.Event  `json:"events"`
}

func (c *Client) DescribePod(ctx context.Context, in DescribePodInput) (*DescribePodOutput, error) {
	pod, err := c.core.CoreV1().Pods(in.Namespace).Get(ctx, in.Name, metav1.GetOptions{})
	if err != nil {
		return nil, err
	}
	events, err := c.GetEvents(ctx, GetEventsInput{Namespace: in.Namespace, InvolvedObject: in.Name})
	if err != nil {
		return nil, err
	}
	return &DescribePodOutput{Pod: pod, Events: events}, nil
}

type GetLogsInput struct {
	Namespace string `json:"namespace"`
	Pod       string `json:"pod"`
	Container string `json:"container,omitempty"`
	Previous  bool   `json:"previous,omitempty"`
	TailLines *int64 `json:"tail_lines,omitempty"`
}

func (c *Client) GetLogs(ctx context.Context, in GetLogsInput) (string, error) {
	tail := int64(200)
	if in.TailLines != nil {
		tail = *in.TailLines
	}
	opts := &corev1.PodLogOptions{
		Container: in.Container,
		Previous:  in.Previous,
		TailLines: &tail,
	}
	req := c.core.CoreV1().Pods(in.Namespace).GetLogs(in.Pod, opts)
	raw, err := req.DoRaw(ctx)
	if err != nil {
		return "", err
	}
	return string(raw), nil
}

type GetEventsInput struct {
	Namespace      string `json:"namespace"`
	InvolvedObject string `json:"involved_object,omitempty"`
}

func (c *Client) GetEvents(ctx context.Context, in GetEventsInput) ([]corev1.Event, error) {
	list, err := c.core.CoreV1().Events(in.Namespace).List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, err
	}
	events := list.Items
	if in.InvolvedObject != "" {
		filtered := events[:0]
		for _, e := range events {
			if e.InvolvedObject.Name == in.InvolvedObject {
				filtered = append(filtered, e)
			}
		}
		events = filtered
	}
	sort.Slice(events, func(i, j int) bool {
		return events[i].LastTimestamp.Time.Before(events[j].LastTimestamp.Time)
	})
	return events, nil
}

type TopPodsInput struct {
	Namespace string `json:"namespace"`
}

func (c *Client) TopPods(ctx context.Context, in TopPodsInput) (any, error) {
	list, err := c.metrics.MetricsV1beta1().PodMetricses(in.Namespace).List(ctx, metav1.ListOptions{})
	if err != nil {
		return map[string]string{"error": "metrics-server unavailable or not installed: " + err.Error()}, nil
	}
	return list.Items, nil
}

File: src/kubewhy/tools.py
"""Read-only Kubernetes tools.

Every function in this module issues exactly one kind of API call: get, list,
watch, or a subresource read (logs). None of them can mutate cluster state.
That's not just a convention here -- kubewhy is meant to be run against a
ServiceAccount whose ClusterRole only grants get/list/watch (see
deploy/readonly-clusterrole.yaml), so even a prompt-injected or hallucinating
model gets a 403 from the API server if it tries anything else.
"""
from __future__ import annotations

from kubernetes import client, config
from kubernetes.client.rest import ApiException

_core: client.CoreV1Api | None = None
_apps: client.AppsV1Api | None = None
_custom: client.CustomObjectsApi | None = None


def load_kube_client() -> None:
    global _core, _apps, _custom
    try:
        config.load_kube_config()
    except config.ConfigException:
        config.load_incluster_config()
    _core = client.CoreV1Api()
    _apps = client.AppsV1Api()
    _custom = client.CustomObjectsApi()


def _fmt_api_error(e: ApiException) -> str:
    return f"API error {e.status}: {e.reason}"


def get_resource(kind: str, namespace: str, name: str | None = None, label_selector: str | None = None) -> dict:
    """Get or list a resource: pod, deployment, replicaset, service, node, event."""
    kind = kind.lower()
    try:
        if kind == "pod":
            if name:
                return _core.read_namespaced_pod(name, namespace).to_dict()
            return [p.to_dict() for p in _core.list_namespaced_pod(namespace, label_selector=label_selector).items]
        if kind == "deployment":
            if name:
                return _apps.read_namespaced_deployment(name, namespace).to_dict()
            return [d.to_dict() for d in _apps.list_namespaced_deployment(namespace, label_selector=label_selector).items]
        if kind == "replicaset":
            return [r.to_dict() for r in _apps.list_namespaced_replica_set(namespace, label_selector=label_selector).items]
        if kind == "service":
            if name:
                return _core.read_namespaced_service(name, namespace).to_dict()
            return [s.to_dict() for s in _core.list_namespaced_service(namespace).items]
        if kind == "node":
            if name:
                return _core.read_node(name).to_dict()
            return [n.to_dict() for n in _core.list_node().items]
        raise ValueError(f"unsupported kind for get_resource: {kind}")
    except ApiException as e:
        return {"error": _fmt_api_error(e)}


def describe_pod(namespace: str, name: str) -> dict:
    """Pod spec/status plus the events involving it -- the read-only equivalent of `kubectl describe pod`."""
    try:
        pod = _core.read_namespaced_pod(name, namespace).to_dict()
    except ApiException as e:
        return {"error": _fmt_api_error(e)}
    events = get_events(namespace, involved_object=name)
    return {"pod": pod, "events": events}


def get_logs(namespace: str, pod: str, container: str | None = None, previous: bool = False, tail_lines: int = 200) -> str:
    try:
        return _core.read_namespaced_pod_log(
            name=pod,
            namespace=namespace,
            container=container,
            previous=previous,
            tail_lines=tail_lines,
        )
    except ApiException as e:
        return _fmt_api_error(e)


def get_events(namespace: str, involved_object: str | None = None) -> list[dict]:
    try:
        events = _core.list_namespaced_event(namespace).items
    except ApiException as e:
        return [{"error": _fmt_api_error(e)}]
    out = [e.to_dict() for e in events]
    if involved_object:
        out = [e for e in out if e.get("involved_object", {}).get("name") == involved_object]
    out.sort(key=lambda e: e.get("last_timestamp") or e.get("event_time") or "")
    return out


def top_pods(namespace: str) -> list[dict] | dict:
    """Requires metrics-server. Read-only -- hits the metrics.k8s.io aggregated API."""
    try:
        result = _custom.list_namespaced_custom_object(
            group="metrics.k8s.io", version="v1beta1", namespace=namespace, plural="pods"
        )
        return result.get("items", [])
    except ApiException as e:
        return {"error": f"metrics-server unavailable or not installed: {_fmt_api_error(e)}"}


def can_i(namespace: str, service_account: str, verb: str, resource: str) -> dict:
    """Read-only RBAC check via SelfSubjectAccessReview-style dry run using rbac.authorization.k8s.io lookups."""
    try:
        roles = _custom.list_namespaced_custom_object(
            group="rbac.authorization.k8s.io", version="v1", namespace=namespace, plural="rolebindings"
        )
        return {"rolebindings": roles.get("items", [])}
    except ApiException as e:
        return {"error": _fmt_api_error(e)}
